### PR TITLE
chore(lspsaga): Migrate to v0.2.3.

### PR DESCRIPTION
**Pull request by ayamir**

### CODE REVIEW

**Comment by ayamir** [comment]
note: click seems not supported.

**Comment by ayamir** [comment]
image
Besides, symbols bg color is not correct currently.

**Comment by Jint-lzxy** [comment]
> Besides, symbols bg color is not correct currently.

Because the highlight groups and methods for color acquisition have changed, this part will be rewritten 👍

**Comment by Jint-lzxy** [comment]
@ayamir Several keybindings have been modified in 0.2.3. Shall we implement support for `Lspsaga show_*_diagnostics` and `Lspsaga *_calls`?

**Comment by ayamir** [comment]
> @ayamir Several keybindings have been modified in 0.2.3. Shall we implement support for `Lspsaga show_*_diagnostics` and `Lspsaga *_calls`?

Of course, but after fixing current config.

**Comment by CharlesChiuGit** [comment]
Currently `Lspsaga outline` can finally work on python, but I will show some errors when using `:q` to close the outline buffer. Not sure it's related to our events or it's an upstream issue.

圖片

**Comment by Jint-lzxy** [comment]
@CharlesChiuGit This is a known issue. I am refactoring all relevant code.

**Comment by CharlesChiuGit** [comment]
I think this PR is ready to go. Basic/Old functionally is pretty much the same.

Just for the note: 
1. lualine hl group issue
2. floating buffer bg improve( I think current bg color of floating buffers don't match catppuccin that well)
3. outline buffer closing errors.
4. new command implementation(`Lspsaga show_*_diagnostics`, `Lspsaga *_calls`)

**Comment by CharlesChiuGit** [comment]
@Jint-lzxy What are u working on? Maybe I can help with other parts?

**Comment by Jint-lzxy** [comment]
@CharlesChiuGit There are still many issues related to highlighting that have not yet been resolved. The first "working commit" will be pushed soon.

> What are u working on? Maybe I can help with other parts?

I am dealing with issues related to highlight groups. Maybe you can help me ask about color replacement of `lspkind` icons in the upstream repository? AFAIU v0.2.3 supports replacing icons but **does not** support replacing the _color_ of those icons like previous versions.

**Comment by CharlesChiuGit** [comment]
U mean changing the color of the icons like in the `Msg` title?

圖片

**Comment by Jint-lzxy** [comment]
Yes. Because using `catppuccin`'s highlight group will cause some bg colors to be inconsistent with lspsaga's default icon colors.

**Comment by CharlesChiuGit** [comment]
https://github.com/glepnir/lspsaga.nvim/commit/ea6acddc9eedd40a1a493a9a06d58e400e1d695e

> 3. outline buffer closing errors.

Seems to be fixed 18 mins age.

**Comment by Jint-lzxy** [comment]
@CharlesChiuGit More specifically, it should be the color of `lspkind` icons. The main problem comes from [this function](https://github.com/glepnir/lspsaga.nvim/blob/f400b54dbfc7a379547cb9cc6f5096a71514f300/lua/lspsaga/lspkind.lua#L5). We can only control the RGB value of the icon's color, but we can't control _which_ color the icon would have.

**Comment by Jint-lzxy** [comment]
This force-push fixes several problems in the config table and resolves the highlight problem of lualine.

**Comment by Jint-lzxy** [comment]
These commits will eventually be squashed into one 👍

**Comment by CharlesChiuGit** [comment]
Seems like we can't do much about changing the icon color right now. I have submitted a feature request, let's see how things go!
https://github.com/glepnir/lspsaga.nvim/issues/669

**Comment by Jint-lzxy** [comment]
> I have submitted a feature request, let's see how things go!

Looks like this issue has been fixed (https://github.com/glepnir/lspsaga.nvim/commit/958e329af091a2843be6c883e057e1aa47460c42). Corresponding changes will be pushed 👍

**Comment by CharlesChiuGit** [comment]
Damn, glepnir is BLAZINGLY FAST lol

**Comment by Jint-lzxy** [comment]
A **breaking change** to `utils` will be pushed later, which would expose a function to get the global color palette. Users can modify specific colors in `settings.lua` or pass in as a function parameter. This feature may require additional debugging.

**Review by CharlesChiuGit — DISMISSED** [review]

File: lua/modules/ui/icons.lua [hunk 1]
```diff
@@ -76,6 +76,7 @@ local data = {
 		BookMark = "",
 		Bug = "",
 		Calendar = "",
+		Card = " ",
```
[the remark this answers is not in the document]
Comment (CharlesChiuGit) [reply lua/modules/ui/icons.lua]: Seems like it's fixed, plz ignore it.

**Comment by Jint-lzxy** [comment]
This commit (https://github.com/ayamir/nvimdots/commit/02c1fd847dd4b9d9dd5061b538a96626b4ea2bcd) defines class `palette`, which is used to manage the global palette.

**Comment by Jint-lzxy** [comment]
The order of priority for modifying the palette is:
preset colors < global colors defined in `settings.lua` < incoming parameters

**Comment by ayamir** [comment]
Approved

**Comment by CharlesChiuGit** [comment]
Jint-lzxy Are u still working on adding custom color to `settings.lua`? Or the PR is ready to go, since it LGTM!

**Comment by Jint-lzxy** [comment]
This is the first draft of https://github.com/ayamir/nvimdots/pull/418#issuecomment-1382690186 :)

**Comment by Jint-lzxy** [comment]
Next, we would need to adjust the colors for `lspsaga` and we're good to go 👍

**Comment by Jint-lzxy** [comment]
This commit added keymaps for the new features of lspsaga. We may need to adjust those keybindings, and then add the entries to the legendary.

**Comment by ayamir** [comment]
> This is the first draft of [#418 (comment)](https://github.com/ayamir/nvimdots/pull/418#issuecomment-1382690186) :) @ayamir @CharlesChiuGit Additional tests may be required. There are two main features to be tested: `palette_overwrite` in `settings.lua` and `require("modules.utils").get_palette([optional table, see desc in lua])`. `vim.inspect` can be used to check whether the results are correct.

Tested for `onedark.vim`, passed.

**Comment by Jint-lzxy** [comment]
> Tested for `onedark.vim`, passed. What's more, no need to `PackerCompile` after set `palette_overwrite`.

Great! Will start rewriting the colors for `lspsaga`. We also need someone to complete the keybinding-related settings _(as well as documenting those in the wiki)_

**Comment by CharlesChiuGit** [comment]
I can setup some keymaps and legendary table first, and see if you guys like it or not.

**Comment by Jint-lzxy** [comment]
> ~I can setup some keymaps and legendary table first, and see if you guys like it or not.~ Seems like it's done lol

@CharlesChiuGit You can add those to the wiki 👍

**Comment by CharlesChiuGit** [comment]
Sure! will do.

**Review by Jint-lzxy — APPROVED** [review]
So far so good 👍 I have resolved conflicts between different highlight groups and squashed the previous patches 🍭

IMO We may use the "square and merge" option to merge this PR due to the large amount of interrelated commits.
merge-options

**Comment by CharlesChiuGit** [comment]
Just out of curiosity, why we need to add filed `NONE`, and why user should never touch that?

**Comment by Jint-lzxy** [comment]
> Just out of curiosity, why we need to add filed `NONE`, and why user should never touch that?

@CharlesChiuGit `"NONE"` implies that this attribute is not set and is usually used to clear the highlight. It is useful in many situations, and we should set this as a default option for the users.

**Comment by CharlesChiuGit** [comment]
tks!

**Comment by CharlesChiuGit** [comment]
I also notice one thing, `colorizer` seems not working before `InsertEnter`. Can u guys confirm that? It seems happens during this rewrite.

**Comment by Jint-lzxy** [comment]
> I also notice one thing, `colorizer` seems not working before `InsertEnter`. Can u guys confirm that? It seems happened during this PR, tho it might not be directly related to this PR.

No, colorizer is always initialized in this way, even before this PR. Otherwise initialization takes a long time.

**Comment by CharlesChiuGit** [comment]
I show @Jint-lzxy 's revision on wiki pages. so fast lol.

**Review by CharlesChiuGit — COMMENTED** [review]
LGTM

**Pull request merged**